enhancements/sig-policy/71-hub-templates-auto-reconcile/README.md

```diff
@@ -0,0 +1,232 @@
+Policy) are watching this object to determine if the watch API request should be canceled after the RemoveWatcher method
+is called.
+
+#### Using The Library
+
+The Policy controller in the Governance Policy Propagator would initially be the only consumer of the library. To use
+it, the controller would essentially have two event loops. The first is the existing reconcile loop that runs on updates
+to Policy and Placement related objects. This event loop would be updated so that any time a root Policy object is
+updated that uses Hub policy templates, dynamic watches would be added, updated, or removed using the new library. To do
+this, the [go-template-utils](https://github.com/stolostron/go-template-utils) library would need to be updated to
+return the objects it queried for on the API. The reconcile loop would then collect references to all objects that were
+queried for using Hub policy templates for that root Policy object and call AddOrUpdateWatcher method from the library.
+When the root Policy object is deleted, the RemoveWatcher method from the library is called.
+
+The second (new) event loop would be the event loop from the new library, where anytime an object referenced in a Hub
+policy template is updated, the root Policy would be reconciled. This would have the same logic as the first event loop,
+and thus this new Reconcile method would just wrap the Reconcile method from the first event loop. Note that since these
+event loops will be run as different Goroutines, locking should be added so that only one root Policy object is
+reconciled at a time.
+
+### Risks and Mitigation
+
+One risk is that creating too many Kubernetes watch API requests could be expensive since we would need to create a
+watch request per object to avoid client-side filtering. Measuring the CPU usage of the Kubernetes API once this is
+implemented is important. If the CPU usage is too high, we could measure the performance of client-side filtering to see
+if it performs better.
+
+In this usage, there will be a limited amount of watches because in practice, many policies don’t use Hub policy
+templates, and when they do, it is usually referencing one object multiple times.
+
+### Open Questions [optional]
+
+N/A
+
+### Test Plan
+
+**Note:** _Section not required until targeted at a release._
+
+Consider the following in developing a test plan for this enhancement:
+
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything that would count as tricky in the
+implementation and anything particularly challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage expectations).
+
+All code is expected to have sufficient e2e tests.
+
+### Graduation Criteria
+
+It would be GA in the release after implementation.
```

Comment:
we should consider performance test before it is GAed.

Reply:
Hi, QE can also plan for testing the specific objects commonly used by customers. Can more info be provided here? re: "...Anything that would count as tricky in the
implementation and anything particularly challenging to test should be called out."?

Also, what is the jira item or git issue for this feature?

Reply:
I agree that performance testing should be done before it can be GA. I think the backup plan of if it's too performance intensive is to make this opt-in.

Reply:
> Hi, QE can also plan for testing the specific objects commonly used by customers. Can more info be provided here? re: "...Anything that would count as tricky in the implementation and anything particularly challenging to test should be called out."?
> 
> Also, what is the jira item or git issue for this feature? Is this for https://issues.redhat.com/browse/ACM-1208?

That is the correct associated Jira.

I would say the most common scenarios are a user copying a secret from the Hub to a managed cluster using the `fromSecret` Hub template function.

Another common scenario is a `ConfigMap` that contains variable configuration (e.g. an operator channel) per managed cluster. For example:

```yaml
apiVersion: v1
kind: ConfigMap
....
data:
  cluster1-gitops-channel: gitops-1.6
  cluster2-gitops-channel: gitops-1.5
  cluster3-gitops-channel: stable
```

Reply:
fyi, @yahliu ⬆️